### ISSUE TITLE
how to pass context and other args in pool_func?

### ISSUE THREAD

**Issue by kungf** [panjf2000/ants#81]
hello, @panjf2000  
`func NewPoolWithFunc(size int, pf func(interface{}), options ...Option)`
in NewPoolWithFunc, pf only have an ingerface{}, but i want to pass at least two args, how to solve tihes?

my run function like :
```
func (s *fileSystemServer) handleOp(
	c *Connection,
	ctx context.Context,
	op interface{}) {
}
```
context is not suggested to put into a struct, how to pass ctx + args as one task to queue?

**Comment by panjf2000** [comment]
You can define a `struct` which contains all your args and use this struct as the single arg of `pf`, then restore it to struct from an interface in `pf` and you can get all args.

@kungf

**Comment by kungf** [comment]
yes, this is what i first want resolution, but  the context arg was not suggested to put into struct, so a little complex 
https://groups.google.com/forum/#!topic/golang-nuts/xRbzq8yzKWI

@panjf2000

**Comment by panjf2000** [comment]
Then I'm out of solutions, maybe I should start considering refactor `pool_func` and make it accept variable arguments like `func NewPoolWithFunc(size int, pf func(...interface{}), options ...Option)`

**Comment by davidxiaozhi** [comment]
@panjf2000 could use this method?
for j := 0; j < batchCount; j++ {
		bj:=j
		routinePoll.Submit(func() {
			genBucketBatch(kv, bucketIndex, bj, pageSize, batchCount, meta)
			atomic.AddInt64(&batchNum, 1)
		})

	}

**Comment by panjf2000** [comment]
I don't quite get your point, what did you want to ask exactly? Is your question related to this issue? if not, please open your own issue.

@davidxiaozhi

**Comment by robin0323** [comment]
> @panjf2000 could use this method?
> 
> ```go
> for j := 0; j < batchCount; j++ {
> 		bj:=j
> 		routinePoll.Submit(func() {
> 			genBucketBatch(bj)
> 		})
>  }
> ```
that works for me, thk~

**Comment by mohammad-babaei** [comment]
> Then I'm out of solutions, maybe I should start considering refactor `pool_func` and make it accept variable arguments like `func NewPoolWithFunc(size int, pf func(...interface{}), options ...Option)`

I have the same problem with args and it would be better if it was ...interface{} instead of interface{}.
Are you gonna fix it?
If not i will have to look for an alternative.
thank you in advance

**Comment by rohmanhm** [comment]
> @panjf2000 could use this method?
> 
> ```go
> for j := 0; j < batchCount; j++ {
> 		bj:=j
> 		routinePoll.Submit(func() {
> 			genBucketBatch(bj)
> 		})
>  }
> ```

Thank you so much! This works for me without creating a new pool with function.

**Comment by Reid00** [comment]
same issue, how to pass multi args of task, is there any good suggestion?

**Comment by tong3jie** [comment]
> 同样的问题，如何传递任务的多个参数，有什么好的建议吗？

```
func A(a string, b string) func() {
	return func() {
		fmt.Println(a, b)
	}
}
f := A("a","b")
pool.Submit(f)

```

**Comment by someview** [comment]
这个问题有进展吗，字节有一个简短的任务池实现，里面也支持ctx。可以提交包含ctx的任务，还是非常重要的，submit(ctx, func())

**Comment by panjf2000** [comment]
> 这个问题有进展吗，字节有一个简短的任务池实现，里面也支持ctx。可以提交包含ctx的任务，还是非常重要的，submit(ctx, func())

传入 context 后，是在用户代码里自己使用还是说要在 `ants` 内部用这个 ctx 控制流程，这个区别很大。
